packages/network/src/connection/WebRtcConnection.ts

```diff
@@ -0,0 +1,402 @@
+let ID = 0
+
+/**
+ * Strict types for EventEmitter interface.
+ */
+interface Events {
+    localDescription: (type: any, description: string) => void
+    localCandidate: (candidate: string, mid: string) => void
+    open: () => void
+    message: (msg: string)  => void
+    close: (err?: Error) => void
+    error: (err: Error) => void
+    bufferLow: () => void
+    bufferHigh: () => void
+}
+
+// reminder: only use Connection emitter for external handlers
+// to make it safe for consumers to call removeAllListeners
+// i.e. no this.on('event')
+export const ConnectionEmitter = EventEmitter as { new(): StrictEventEmitter<EventEmitter, Events> }
+
+export class DeferredConnectionAttempt {
+    
+    private eventEmitter: EventEmitter
+    private connectionAttemptPromise: Promise<string>
+
+    constructor(private targetId: string) {
+
+        this.eventEmitter = new EventEmitter()
+        
+        this.connectionAttemptPromise = new Promise((resolve, reject) => {
+            this.eventEmitter.once('resolve', () => {
+                resolve(this.targetId)
+            })
+            this.eventEmitter.once('reject', (reason) => {
+                reject(reason)
+            })
+        })
+
+        // allow promise to reject without outside catch
+        this.connectionAttemptPromise.catch(() => {})
+    }
+
+    getPromise(): Promise<string> {
+        return this.connectionAttemptPromise
+    }
+
+    resolve(): void {
+        this.eventEmitter.emit('resolve')
+    }
+
+    reject(reason: Error | string): void {
+        this.eventEmitter.emit('reject', reason)
+    }
+}
+
+export function isOffering(myId: string, theirId: string): boolean {
+    return myId < theirId
+}
+
+export abstract class WebRtcConnection extends ConnectionEmitter {
+    public readonly id: string
+    private connectionId = 'none'
+    private peerInfo: PeerInfo
+    private readonly routerId: string
+    
+    private readonly maxPingPongAttempts: number
+    private readonly pingInterval: number
+    private readonly flushRetryTimeout: number
+    private readonly messageQueue: MessageQueue<string>
+   
+    private flushRef: NodeJS.Immediate | null
+    private flushTimeoutRef: NodeJS.Timeout | null
+
+    private pingTimeoutRef: NodeJS.Timeout | null
+    private deferredConnectionAttempt: DeferredConnectionAttempt | null
+
+    protected readonly logger: Logger
+    protected readonly maxMessageSize: number
+    protected isFinished: boolean
+    protected readonly selfId: string
+    protected readonly stunUrls: string[]
+    protected connectionTimeoutRef: NodeJS.Timeout | null
+    protected readonly newConnectionTimeout: number 
+    protected lastState: string | null
+    protected lastGatheringState: string | null
+    protected paused: boolean
+    protected readonly bufferThresholdHigh: number
+    protected readonly bufferThresholdLow: number
+    protected pingAttempts = 0
+    protected rtt: number | null
+    protected rttStart: number | null
```

Comment:
Let's try to make as many as possible to `protected readonly`, if not possible keep `private` and provide getter.